### ISSUE TITLE
Mypy doesn't detect self when binding methods

### ISSUE THREAD

**Issue by jfuruness** [python/mypy#17737]
**Bug Report**

If you create a function and then bind that to a class, subclasses of the parent class will not detect the self parameter properly

**To Reproduce**

```python
def func(self) -> None:
    print("in parent")

class Parent:
    func = func

class Child(Parent):
    def func(self) -> None:
        print("in child")
```

**Expected Behavior**

All checks should pass, and when you assign func in the Parent class, the self type should no longer be Any

**Actual Behavior**

```
test.py:8: error: Signature of "func" incompatible with supertype "Parent"  [override]
test.py:8: note:      Superclass:
test.py:8: note:          def func(self: Any) -> None
test.py:8: note:      Subclass:
test.py:8: note:          def func(self) -> None
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

- Mypy version used: 1.11.2
- Mypy command-line flags: none
- Mypy configuration options from `mypy.ini` (and other config files): none
- Python version used: 3.12.3
